frontend/src/common/hooks/useCategoryFilter/useCategoryFilter.ts

```diff
@@ -0,0 +1,897 @@
+          labelsById.set(ontology_term_id, label);
+        }
+      }
+    );
+  });
+}
+
+/**
+ * Build map of ontology term labels keyed by ontology term IDs. This map is used to determine the display values for
+ * ontology term-backed fields (e.g. tissue).
+ * @param originalRows - Original result set before filtering.
+ * @param categoryFilterIds - Set of category filter IDs to include for this filter instance.
+ * @returns Map of ontology term labels keyed by ID.
+ */
+function keyOntologyTermLabelsById<T extends Categories>(
+  originalRows: Row<T>[],
+  categoryFilterIds: Set<CATEGORY_FILTER_ID>
+): Map<string, string> {
+  const labelsById = new Map<string, string>();
+
+  // Collect the set of term sets across all category filters.
+  const termSets = [...categoryFilterIds].reduce(
+    (accum: OntologyTermSet[], categoryFilterId: CATEGORY_FILTER_ID) => {
+      const config = CATEGORY_FILTER_CONFIGS_BY_ID[categoryFilterId];
+
+      // Add source of curated ontology categories.
+      if (config.valueSourceKind === "CURATED") {
+        accum.push(config.source);
+        return accum;
+      }
+
+      // Add source for each panel of multi-panel categories.
+      if (config.viewKind === "MULTI_PANEL") {
+        config.panels.forEach((panel: CategoryFilterPanelConfig) => {
+          if (panel.sourceKind === "CURATED") {
+            accum.push(panel.source);
+          }
+        });
+      }
+      return accum;
+    },
+    [] as OntologyTermSet[]
+  );
+
+  // Key curated term sets.
+  termSets.forEach((termSet: OntologyTermSet) => {
+    keyCuratedOntologyTermLabelsById(labelsById, termSet);
+  });
+
+  // Key ontology terms from row values.
+  ["cell_type", "tissue"].forEach((filterOnKey) => {
+    keyRowOntologyTermLabelsById(
+      labelsById,
+      originalRows,
+      filterOnKey as FilterKey
+    );
+  });
+
+  return labelsById;
+}
+
+/**
+ * Update selected state of categories to match the current set of selected filters.
+ * @param nextFilterState - Filter state being built on select of filter.
+ * @param filters - Current set of selected category values (values) or ranges keyed by category (id).
+ * @param multiPanelUIState - Current set of category values that the user has selected on the UI for all multi-
+ * panel category filters.
+ */
+function setSelectedStates<T extends Categories>(
+  nextFilterState: FilterState,
+  filters: Filters<T>,
+  multiPanelUIState: MultiPanelUIState
+) {
+  Object.keys(nextFilterState).forEach((strCategoryFilterId: string) => {
+    const categoryFilterId = strCategoryFilterId as CATEGORY_FILTER_ID;
+
+    // Grab the filter state for this category.
+    const categoryFilterState = nextFilterState[categoryFilterId];
+
+    // Grab the filters for this category, used to determine the selected values.
+    const config = CATEGORY_FILTER_CONFIGS_BY_ID[categoryFilterId];
+    const isMultiPanel = isMultiPanelCategoryFilterConfig(config);
+    const categoryFilter = isMultiPanel
```

Comment:
Selected state is build from multi-panel UI state for multi-panel category filters.